### PR TITLE
Add default imagePullPolicy for initContainers.

### LINKED ISSUE #38542: imagePullPolicy is required on Init Containers within a Pod template
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):

No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

* initContainers imagePullPolicy
* initContainers required value

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

BUG REPORT

**Kubernetes version** (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"5+", GitVersion:"v1.5.0-beta.3", GitCommit:"35fd1363821b8432ccde55632cf9ead4c79eddbb", GitTreeState:"clean", BuildDate:"2016-12-08T21:58:56Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5+", GitVersion:"v1.5.0-beta.3", GitCommit:"35fd1363821b8432ccde55632cf9ead4c79eddbb", GitTreeState:"clean", BuildDate:"2016-12-08T21:54:19Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: GCE
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**: cluster/kube-up.sh
- **Others**:

**What happened**:

When trying to create either a ReplicaSet or Replication Controller whose Pod template contains an Init Container, the validation fails saying `imagePullPolicy` is a required value.

```
The ReplicaSet "nginx" is invalid: spec.template.spec.initContainers[0].imagePullPolicy: Required value
```

```
The ReplicationController "nginx" is invalid: spec.template.spec.initContainers[0].imagePullPolicy: Required value
```

**What you expected to happen**:

The create should succeed and the Init Container should receive a default `imagePullPolicy`.

**How to reproduce it** (as minimally and precisely as possible):

Try to create either of the following:

**ReplicaSet**

```yaml
apiVersion: extensions/v1beta1
kind: ReplicaSet
metadata:
  name: nginx
spec:
  replicas: 3
  template:
    metadata:
      labels:
        app: nginx
      annotations:
        pod.beta.kubernetes.io/init-containers: '[
            {
                "name": "install",
                "image": "busybox",
                "command": ["wget", "-O", "/work-dir/index.html", "http://kubernetes.io/index.html"],
                "volumeMounts": [
                    {
                        "name": "workdir",
                        "mountPath": "/work-dir"
                    }
                ]
            }
        ]'
    spec:
      containers:
      - name: nginx
        image: nginx
        ports:
        - containerPort: 80
        volumeMounts:
        - name: workdir
          mountPath: /usr/share/nginx/html
      dnsPolicy: Default
      volumes:
      - name: workdir
        emptyDir: {}
```

**Replication Controller**

```yaml
apiVersion: v1
kind: ReplicationController
metadata:
  name: nginx
spec:
  replicas: 3
  selector:
    app: nginx
  template:
    metadata:
      labels:
        app: nginx
      annotations:
        pod.beta.kubernetes.io/init-containers: '[
            {
                "name": "install",
                "image": "busybox",
                "command": ["wget", "-O", "/work-dir/index.html", "http://kubernetes.io/index.html"],
                "volumeMounts": [
                    {
                        "name": "workdir",
                        "mountPath": "/work-dir"
                    }
                ]
            }
        ]'
    spec:
      containers:
      - name: nginx
        image: nginx
        ports:
        - containerPort: 80
        volumeMounts:
        - name: workdir
          mountPath: /usr/share/nginx/html
      dnsPolicy: Default
      volumes:
      - name: workdir
        emptyDir: {}
```

**Anything else do we need to know**:

If I add `"imagePullPolicy": "IfNotPresent",` to the Init Container in the ReplicaSet or Replication Controller Pod template, the create succeeds.

When I create an Init Container without `imagePullPolicy` in a StatefulSet Pod template, it succeeds.

When I create an Init Container without `imagePullPolicy` in a bare Pod, it succeeds:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx
  annotations:
    pod.beta.kubernetes.io/init-containers: '[
        {
            "name": "install",
            "image": "busybox",
            "command": ["wget", "-O", "/work-dir/index.html", "http://kubernetes.io/index.html"],
            "volumeMounts": [
                {
                    "name": "workdir",
                    "mountPath": "/work-dir"
                }
            ]
        }
    ]'
spec:
  containers:
  - name: nginx
    image: nginx
    ports:
    - containerPort: 80
    volumeMounts:
    - name: workdir
      mountPath: /usr/share/nginx/html
  dnsPolicy: Default
  volumes:
  - name: workdir
    emptyDir: {}
```

### PR DESCRIPTION
**Pull request by k82cn**
fixes #38542

### CODE REVIEW

**Comment by k82cn** [comment]
I'm going to append test result later.

**Comment by k82cn** [comment]
Test Cases: `ReplicaSet`'s initContainers' without `imagePullPolicy ` (refer to yml in #38542)
Test Steps:

Before:
```
$kubectl create -f 38542.yml
The ReplicaSet "nginx" is invalid: spec.template.spec.initContainers[0].imagePullPolicy: Required value
```

After:
```
$kubectl create -f 38542.yml
replicaset "nginx" created
```

**Comment by calebamiles** [comment]
@k8s-bot cri e2e test this

**Comment by foxish** [comment]
This should be in the 1.5 milestone and cherrypicked I imagine.

cc @dims @saad-ali

**Comment by k82cn** [comment]
@k8s-bot cri e2e test this

**Comment by kingdonb** [comment]
Anyone know of a reason this hasn't been merged ?

**Comment by kingdonb** [comment]
Not sure if Jenkins CRI GCE e2e is broken upstream, I don't think the failure is as a result of this change though.

**Comment by k82cn** [comment]
@k8s-bot cri e2e test this

**Comment by k82cn** [comment]
@k8s-bot cri e2e test this

File: pkg/api/v1/conversion.go [hunk 1]
```diff
@@ -493,6 +493,12 @@ func Convert_v1_PodTemplateSpec_To_api_PodTemplateSpec(in *PodTemplateSpec, out
 		// taking responsibility to ensure mutation of in is not exposed
 		// back to the caller.
 		in.Spec.InitContainers = values
+
+		// Call defaulters explicitly until annotations are removed
+		for i := range in.Spec.InitContainers {
```
Comment (smarterclayton) [inline pkg/api/v1/conversion.go]: This doesn't get called by the kubelet - are the defaults being correctly applied there?

**Comment by smarterclayton** [comment]
Please add a test case that demonstrates this.

**Comment by k82cn** [comment]
Sure, I'll add a test case for that :).

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/api/v1/conversion.go, line 498 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/38574#-KYgymzoD0pgFvRGOVWF-r1-498:-KZTgSGzpWYsar5RKcCH:b-2ppusm) ([raw file](https://github.com/kubernetes/kubernetes/blob/11e5da57f6675dce5e7939c451447c75f58ed274/pkg/api/v1/conversion.go#L498)):*
<details><summary><i>Previously, smarterclayton (Clayton Coleman) wrote…</i></summary><blockquote>

Never mind - kubelet is non-defaulting now.

This won't work in all possible code paths in the api server, but is probably sufficient for now.
</blockquote></details>

After defaulting in apiserver, the kubelet will get it in serialize/de-serialize.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/38574#-:-KZThD0ulmIMq86ee6ba:bb3vwc1)*

**Comment by k82cn** [comment]
@k8s-bot verify test this

**Comment by k82cn** [comment]
Added a test case for this issue.

/cc @smarterclayton

**Comment by k82cn** [comment]
@smarterclayton , PTAL :).

**Comment by smarterclayton** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by k82cn** [comment]
@k8s-bot bazel test this

**Pull request merged**